portalnet/src/gossip.rs

```diff
@@ -135,6 +114,151 @@ pub fn propagate_gossip_cross_thread<TContentKey: OverlayContentKey>(
+                    .collect::<Vec<&kbucket::Node<NodeId, Node>>>()
+            })
+            .collect();
+
+        if all_nodes.is_empty() {
+            warn!("No connected nodes, using disconnected nodes for gossip.");
+            all_nodes = kbuckets
+                .buckets_iter()
+                .flat_map(|kbucket| {
+                    kbucket
+                        .iter()
+                        .collect::<Vec<&kbucket::Node<NodeId, Node>>>()
+                })
+                .collect();
+        }
+
+        if all_nodes.is_empty() {
+            // If there are no nodes whatsoever in the routing table the gossip cannot proceed.
+            warn!("No nodes in routing table, gossip cannot proceed.");
+            return gossip_result;
+        }
+        calculate_interested_enrs(&content_key, &all_nodes)
+    };
+    if interested_enrs.is_empty() {
+        return gossip_result;
+    };
+
+    // Create and send OFFER overlay request to the interested nodes
+    for enr in interested_enrs.into_iter() {
+        let (result_tx, mut result_rx) = tokio::sync::mpsc::unbounded_channel();
+        let offer_request = Request::PopulatedOfferWithResult(PopulatedOfferWithResult {
+            content_item: (content_key.clone().into(), data.clone()),
+            result_tx,
+        });
+
+        let (tx, rx) = oneshot::channel();
+        let responder = Some(tx);
+        let overlay_request = OverlayRequest::new(
+            offer_request,
+            RequestDirection::Outgoing {
+                destination: enr.clone(),
+            },
+            responder,
+            None,
+        );
+        // mark peer as offered in all cases except where offer for content key was accepted
+        if let Err(err) = command_tx.send(OverlayCommand::Request(overlay_request)) {
+            error!(error = %err, "Error sending OFFER message to service");
+            gossip_result.offered.push(enr);
+            continue;
+        }
+        match rx.await {
+            Ok(res) => {
+                if let Ok(Response::Accept(accept)) = res {
+                    if accept.content_keys.is_zero() {
+                        gossip_result.offered.push(enr);
+                        continue;
+                    }
+                } else {
+                    gossip_result.offered.push(enr);
+                    continue;
+                }
+            }
+            Err(_) => {
+                gossip_result.offered.push(enr);
+                continue;
+            }
+        }
+        // mark peer as accepted if something went wrong during utp transfer
+        match result_rx.recv().await {
+            Some(result) => {
+                if result {
+                    gossip_result.transferred.push(enr);
+                } else {
+                    gossip_result.accepted.push(enr);
+                }
+            }
+            None => gossip_result.accepted.push(enr),
```

Comment:
I think a lot of this is simplified be having every peer that is offered the content get added to the list, etc.